src/doorman/Channel.cpp

```diff
@@ -0,0 +1,393 @@
+                    break;
+                default: SendCom(fd_, DONT, *(ptr + 2)); break;
+                }
+                memmove(ptr, ptr + 3, nLeft - 3);
+                nLeft -= 3;
+                break;
+            case WONT:
+                SendCom(fd_, DONT, *(ptr + 2));
+                memmove(ptr, ptr + 3, nLeft - 3);
+                nLeft -= 3;
+                break;
+            case DO:
+                if (ptr[2] == TELOPT_ECHO && echoing_ == false) {
+                    SendCom(fd_, WILL, TELOPT_ECHO);
+                    memmove(ptr, ptr + 3, nLeft - 3);
+                    nLeft -= 3;
+                } else {
+                    SendCom(fd_, WONT, *(ptr + 2));
+                    memmove(ptr, ptr + 3, nLeft - 3);
+                    nLeft -= 3;
+                }
+                break;
+            case DONT:
+                SendCom(fd_, WONT, *(ptr + 2));
+                memmove(ptr, ptr + 3, nLeft - 3);
+                nLeft -= 3;
+                break;
+            case SB: {
+                char sbType;
+                byte *eob, *p;
+                /* Enough room for an SB command? */
+                if (nLeft < 6) {
+                    nLeft = 0;
+                    break;
+                }
+                sbType = *(ptr + 2);
+                //          sbWhat = *(ptr + 3);
+                /* Now read up to the IAC SE */
+                eob = ptr + nLeft - 1;
+                for (p = ptr + 4; p < eob; ++p)
+                    if (*p == IAC && *(p + 1) == SE)
+                        break;
+                /* Found IAC SE? */
+                if (p == eob) {
+                    /* No: skip it all */
+                    nLeft = 0;
+                    break;
+                }
+                *p = '\0';
+                /* Now to decide what to do with this new data */
+                switch (sbType) {
+                case TELOPT_TTYPE:
+                    log_out("[%d] TTYPE = %s", id_, ptr + 4);
+                    if (SupportsAnsi((const char *)ptr + 4))
+                        ansi_ = true;
+                    break;
+                case TELOPT_NAWS:
+                    width_ = ptr[4];
+                    height_ = ptr[6];
+                    log_out("[%d] NAWS = %dx%d", id_, width_, height_);
+                    break;
+                }
+
+                /* Remember eob is 1 byte behind the end of buffer */
+                memmove(ptr, p + 2, nLeft - ((p + 2) - ptr));
+                nLeft -= ((p + 2) - ptr);
+
+                break;
+            }
+
+            default:
+                memmove(ptr, ptr + 2, nLeft - 2);
+                nLeft -= 2;
+                break;
+            }
+
+        } else {
+            ptr++;
+            nRealBytes++;
+            nLeft--;
+        }
+    }
+    /* Now to update the buffer stats */
+    buffer_.resize(nRealBytes);
+
+    /* Second pass - look for whole lines of text */
```

Comment:
Yeah surely there's a nice modern way to do a string split on newlines :)

Reply:
You'd think. But yeah :) We can do a lot better!